### ISSUE TITLE
support arm64 architecture

### ISSUE THREAD

**Issue by longquan7** [docker-library/mysql#318]
I want to build this image on arm64 architecture but I don't find arm64 package in this file http://repo.mysql.com/apt/debian/dists/jessie/InRelease , So I want to know the image whether or not support arm64 architecture.

**Comment by yosifkit** [comment]
This would need support from MySQL upstream to release packages for other architectures in their apt repo. (to me, this seems unlikely since they don't commercially support much more then i386 and amd64: https://www.mysql.com/support/supportedplatforms/database.html)

cc @ltangvald would mysql be willing to build their apt packages on more architectures? (like arm64v8, arm32v7, ppc64le, and s390x)

**Comment by gnumoksha** [comment]
@yosifkit  [debian has mysql server package](https://packages.debian.org/search?suite=stretch&arch=arm&searchon=names&keywords=mysql) for arm architecture. I am using it on a few Raspberry Pi.

**Comment by tianon** [comment]
Debian's versions won't be up to date with upstream's, which is why these
images consume packages exclusively from upstream's repositories.

In the meantime, I'd recommend creating your own image which installs
Debian's packages if this is something you need right away.

**Comment by smallfryy** [comment]
@longquan7 - did you find a solution? I'm also looking to build mySQL:5.7 image on arm64 architecture (Odroid XU3, per issue referenced above).

**Comment by longquan7** [comment]
@samalba Nothing yet.This would need support from MySQL upstream to release packages.

**Comment by geerlingguy** [comment]
Just a note for anyone who needs a working image for testing purposes (though I'd recommend building your own for prod!) for Raspberry Pi at least—I've tested the following two and can confirm they work okay on my Pi 3 B and 3 B+ with Raspbian:

  - https://hub.docker.com/r/hypriot/rpi-mysql/
  - https://hub.docker.com/r/tobi312/rpi-mysql/

Note that they only support MySQL 5.5 at this time.

**Comment by shawnguo2** [comment]
@ltangvald I'm particularly wondering about mysql's take on arm64 architecture support, considering there are a lot of push on arm64 enterprise from industry.  Is there any plan already being discussed or just no, nothing on that?  Thanks much.

**Comment by ltangvald** [comment]
We do have an experimental preview out https://blogs.oracle.com/wim/mysql-805-community-edition-preview-for-oracle-linux-7-for-arm64-preview (8.0.5 was never released, but was instead renamed to 8.0.11), but I don't really have any more concrete information than that :)

Debian stretch does not have MySQL. They just have a mysql-server package that actually points to MariaDB. Ubuntu has MySQL, but generally only a single version (5.7)

5.7 and 8.0 can also be built on raspberry pi, but I think the build needs to be tweaked since the device is short on memory.

**Comment by shawnguo2** [comment]
@ltangvald Thanks much for your input.  It's very helpful for us to understand the status of MySQL on arm64.  As per comments from @yosifkit and @tianon, we should only get MySQL package from MySQL upstream repo.  Before MySQL releases debian package for arm64, It seems the only way for getting arm64 support in official image is to add an OL (Oracle Linux) variant for MySQL docker image?  Or is there any other way around?

**Comment by ltangvald** [comment]
The only other option would be to build the server from scratch, on arm64, and it's unlikely to be made a part of the official image.

Regarding an OL variant, we're hoping to add them to the official image soon (this is hanging on me, as I need to make a templating system so we don't double the maintenance work by adding more images). 

We do already have the OL variant upstream (mysql/mysql-server instead of just mysql), but that doesn't have an arm64 version either.

**Comment by ltangvald** [comment]
The currently available MySQL package for arm64 was an experimental, one-off release. We wouldn't be able to add it to the mysql/mysql-server image until it's supported for the regular releases (in which case we'll most likely have access to the same minimal-server images).

**Comment by lag-linaro** [comment]
Out of interest, what are the likely timescales for having ARM64 support for the regular releases?

Is there anything we can do to expedite the process?

**Comment by tianon** [comment]
@ltangvald it appears that https://repo.mysql.com/yum/mysql-8.0-community/docker/aarch64/ is looking a bit more official now :smirk: 

Any chance that's coming to the APT builds soon, and/or that `arm64v8` support for `mysql/mysql-server` is on the horizon? :smile: :+1:

**Comment by ltangvald** [comment]
Hi,

For the former (arm support for apt platforms), I unfortunately don't know.
For the latter, I'm actually working on it right now (I basically just need to figure out the multi-arch manifests and it will be published).
Once the changes I proposed in pr #471 have been finalized (as I mention there it might be more sensible to split it up) I was going to make another one to add rpm-based images to the official repo, for 5.7 and 8.0.

**Comment by ltangvald** [comment]
Hi,

We've publish the upstream aarch64 image now. Since it's still somewhat experimental we used a separate tag: mysql/mysql-server:8.0-aarch64

**Comment by lag-linaro** [comment]
Do you have a link to the PR please?

**Comment by ltangvald** [comment]
It's not properly added to our build pipeline yet, but the source for the arm image can be seen here: https://github.com/ltangvald/mysql-1/blob/mysql-server-arm/8.0-aarch64/Dockerfile

The only thing changed from https://github.com/mysql/mysql-docker/blob/mysql-server/8.0/Dockerfile is the rpm url, plus a required workaround noted at https://dev.mysql.com/doc/refman/8.0/en/linux-installation-yum-repo.html#yum-install-platform-specifics (at line 21 and 24)

**Comment by lag-linaro** [comment]
It's been a while, so I thought I'd check in.  I see that ARM64v8 is still not supported.

Have you guys moved forward in any way?

**Comment by keithy** [comment]
Not sure what the big deal is, there are plenty of builds of MySQL and MariaDB out there for arm64v8. Cant xwiki just be published and we pick our own DB backend?

Also see https://blog.hypriot.com/post/setup-simple-ci-pipeline-for-arm-images/

**Comment by lag-linaro** [comment]
Sorry to keep bugging you about this every 2 months (:wink:), but are we any closer to solving this?

I see #471 gained quite a lot of traction, but it seems to have stagnated.

**Comment by lag-linaro** [comment]
So I tried to open a discussion on AArch64 releases on https://forums.mysql.com, but my post was **unapproved without explanation** which was frustrating.  Especially seeing as it's the only official platform to request support without "contacting a sales representative".

@ltangvald do you know why the AArch64 package(s) has been released for YUM rather than APT.  The x84_64 packages seem to have been released for the latter.

Also, could someone tell me what the difference between this and https://hub.docker.com/r/mysql/mysql-server is please?  The latter already seems to have AArch64 support and is subsequently `docker pull`able.

**Comment by ltangvald** [comment]
AArch64 support is in its early stages for MySQL, but I don't know what the plans are for future releases with regards to platform support.

Essentially, mysql/mysql-server is the image that is made and owned by Oracle (there are also mysql/mysql-cluster and mysql/mysql-router images). It's made to have the same basic functionality, but is missing certain things like the gosu support. The big technical difference is that it uses oraclelinux:7-slim as the base image, so it can use the existing aarch64 packages published by Oracle. It also uses an rpm custom-made for docker, which is stripped down, while the official mysql image uses our regular apt packages (but for 8.0 those provide -core packages that are similarly stripped).

**Comment by lag-linaro** [comment]
@ltangvald thanks for your reply, I appreciate it.

So that's why the AArch64 package at https://repo.mysql.com/yum/mysql-8.0-community/docker/aarch64 uses YUM instead of APT, so it can be consumed with Oracle Linux which is RPM based?

Hopefully your [Templatize and refactor 5.7+ entrypoint scripts](https://github.com/docker-library/mysql/pull/471) PR can be merged soon and you can work on supporting the AArch64 releases via this project.  It would be great to have `docker pull mysql` working on AArch64 platforms.

Do you mind me asking what the implementation details would look like?  I assume, since they use a different package manager a new `Dockerfile` will be required?

**Comment by acodercat** [comment]
hope arm image

**Comment by aarch64-laptops** [comment]
Still blocked on: https://github.com/docker-library/mysql/pull/471

**Comment by rhenwood-arm** [comment]
Is there any news on the progress of this AArch64 image?

**Comment by lag-linaro** [comment]
Looks like the dependency is still not satisfied: https://github.com/docker-library/mysql/pull/471

**Comment by everflux** [comment]
I would like to see an arm image as well. For the mean time I will use the official mariadb arm image https://hub.docker.com/_/mariadb perhaps it is sufficient.

**Comment by lag-linaro** [comment]
Is #471 really still blocking this issue?

This ticket has been open a year now - it seems that people are fast moving to MariaDB.

Is there anything else we can do to unblock this?

**Comment by lag-linaro** [comment]
Looks like I missed the 2 year anniversary! :birthday:

Any news good folks?

**Comment by lag-linaro** [comment]
I'm guessing you know where this is going ...

Now that #471 has been merged, could we turn our attention to supporting `aarch64` please?

**Comment by ltangvald** [comment]
Yeah. My idea for this is to add support for rpm-based images since repo.mysql.com offers rpms for aarch64. This would only be for 8.0, though.

**Comment by lag-linaro** [comment]
What would happen for subsequent releases?

**Comment by ltangvald** [comment]
You mean when e.g. there's a 9.0? arm support for 8.0 wouldn't be removed (it's just that 8.0 was the first version upstream added official arm support for).

**Comment by arjunk-ampere** [comment]
Hi,

Pinging on this issue.  A few questions as well, hopefully not too elementary:
1.  Should we be expecting an 8.x ARM64 mysql official image, or is it delayed til 9.0?
2.  If I search on "mysql-server", I see that Oracle has ARM64 containers for mysql.  These aren't labelled official, and not sure (outside of supported architectures) what the deltas would be between those containers and containers here.

Thanks in advance,
Arjun

**Comment by geerlingguy** [comment]
It seems like the arm64 images are built now, for 8.0 only, for the mysql/mysql-server image at least: https://hub.docker.com/r/mysql/mysql-server/tags

**Comment by Code-Hex** [comment]
Hey, if you want to run MySQL 5.7, let's use this one
https://hub.docker.com/r/codehex/ubuntu-mysql

**Comment by tisba** [comment]
> It seems like the arm64 images are built now, for 8.0 only, for the mysql/mysql-server image at least: https://hub.docker.com/r/mysql/mysql-server/tags

It doesn't look like they are getting updated regularly. Latest 8.x are not available in arm64 :(

**Comment by ltangvald** [comment]
Which image are you missing? 8.0.22 should be available for arm.

**Comment by rfay** [comment]
library/mysql does *not* have arm64 images for mysql 8. See https://hub.docker.com/_/mysql?tab=tags&page=1&ordering=last_updated

library/mysql-server *does* have built arm64 images, but it's a completely different an unrelated build (Not debian). https://hub.docker.com/r/mysql/mysql-server/tags?page=1&ordering=last_updated

**Comment by ltangvald** [comment]
Right, because MySQL upstream does not support|publish arm packages on Debian, only oraclelinux (which is what mysql/mysql-server is based on).

**Comment by hiroshi** [comment]
FYI:
If you just want to run mysql image on docker Desktop for M1 mac, you can use amd64 image (with QUEM I think).

docker-compose.yml
```
services:
  mysql:
    image: mysql:5
    platform: linux/amd64
    ...
```
It works for me.

(or `docker (pull|run) --platform=linux/amd64` maybe)

**Comment by TheHuge** [comment]
Official Docker's documentation mentions this and also proposes tu use **mariadb** ARM64 image as a workaround: [https://docs.docker.com/docker-for-mac/apple-silicon/#known-issues](url)

**Comment by rfay** [comment]
MariaDB is a fine workaround for some uses, but MySQL 5.7 and then extensively 8.0 have moved quite a ways from their roots. So for many uses, MariaDB is not a reasonable workaround.

**Comment by hiroshi** [comment]
@TheHuge Hi, just give you head up. Your link is broken, but thanks for the information 👍 
<a href="https://gyazo.com/f715b633ca6b76b2ecfdf7ff2ad3d872"><img src="https://i.gyazo.com/f715b633ca6b76b2ecfdf7ff2ad3d872.png" alt="Image from Gyazo" width="928"/></a>

**Comment by khalwat** [comment]
Hoping for ARM support for Apple Silicon M1, etc. in the official Docker images, using `mysql/mysql-server` for now instead.

Thanks for all you do!

**Comment by ehsanagh** [comment]
And now I hit this problem on my Apple M1 chip laptop.
Seriously any further plan to make this work?

**Comment by jmcollin78** [comment]
@hiroshi when I try your recommandation, I always fall into this error:

```
runtime: failed to create new OS thread (have 2 already; errno=22)
fatal error: newosproc

runtime stack:
runtime.throw(0x524da0, 0x9)
	/usr/local/go/src/runtime/panic.go:527 +0x90
runtime.newosproc(0xc820028000, 0xc820037fc0)
	/usr/local/go/src/runtime/os1_linux.go:150 +0x1ab
runtime.newm(0x555ce8, 0x0)
	/usr/local/go/src/runtime/proc1.go:1105 +0x130
runtime.main.func1()
	/usr/local/go/src/runtime/proc.go:48 +0x2c
runtime.systemstack(0x5c4300)
	/usr/local/go/src/runtime/asm_amd64.s:262 +0x79
runtime.mstart()
	/usr/local/go/src/runtime/proc1.go:674

goroutine 1 [running]:
runtime.systemstack_switch()
	/usr/local/go/src/runtime/asm_amd64.s:216 fp=0xc820022770 sp=0xc820022768
runtime.main()
	/usr/local/go/src/runtime/proc.go:49 +0x62 fp=0xc8200227c0 sp=0xc820022770
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1 fp=0xc8200227c8 sp=0xc8200227c0

```
My MySQL docker commands are the following:

```
$ docker pull --platform linux/amd64 mysql:5.7.26
...
Status: Downloaded newer image for mysql:5.7.26
docker.io/library/mysql:5.7.26
$  docker run --platform linux/amd64 --name MySQL -p 3306:3306 mysql:5.7.26
runtime: failed to create new OS thread (have 2 already; errno=22)
fatal error: newosproc

runtime stack:
runtime.throw(0x524da0, 0x9)
	/usr/local/go/src/runtime/panic.go:527 +0x90
runtime.newosproc(0xc820028000, 0xc820037fc0)
	/usr/local/go/src/runtime/os1_linux.go:150 +0x1ab
runtime.newm(0x555ce8, 0x0)
	/usr/local/go/src/runtime/proc1.go:1105 +0x130
runtime.main.func1()
	/usr/local/go/src/runtime/proc.go:48 +0x2c
runtime.systemstack(0x5c4300)
	/usr/local/go/src/runtime/asm_amd64.s:262 +0x79
runtime.mstart()
	/usr/local/go/src/runtime/proc1.go:674

goroutine 1 [running]:
runtime.systemstack_switch()
	/usr/local/go/src/runtime/asm_amd64.s:216 fp=0xc820022770 sp=0xc820022768
runtime.main()
	/usr/local/go/src/runtime/proc.go:49 +0x62 fp=0xc8200227c0 sp=0xc820022770
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1 fp=0xc8200227c8 sp=0xc8200227c0
```

Do I miss something ?

I'm on MacOS Big Sure on Mac mini M1.
```
 $ uname -a
Darwin mac-mini.home 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 x86_64
```

**Comment by gtothill** [comment]
Hi all,
Forgive me if Ive missed something here. I'm running the alpha Rasbian (Raspberry OS) Buster arm64 build on a raspberry Pi 4.
Id like to run xwiki on it and move it off my server in an effort to reduce noise - it sounds like an aircraft taking off with its cooling fans!).
I'm running docker and mysql/mysql-server:8.0.25-1.2.3-server arm64/v8 in a container. It works fine (https://hub.docker.com/r/mysql/mysql-server/tags?page=1&ordering=last_updated).
I know that there are lots of postgres images available (https://hub.docker.com/r/arm64v8/xwiki/tags?page=1&ordering=last_updated) but as has been pointed out, no images for mysql.
Id like to be able to use mysql for consistency and because Im hosting other databases in that container so it seems sensible.
I may have missed the point but given there is a build of mysql that seems to work fine on Rasbian arm64.v8, is there any difficulty in producing that image?
Would it be a matter of suing one of the postgres images and simply point it at a mysql database?
Any help or guidance would be most appreciated.

Thanks,

Geoff.

**Comment by gtothill** [comment]
So sorry - I posted this comment in the wrong repo. Should have been in the xwiki repo. Please ignore - I understand the puzzled emoji now tianon. Cheers.

**Comment by rfay** [comment]
I note that Ubuntu 20.04 (arm64) has mysql-server-8.0 and Ubuntu 18.04 (arm64) has mysql-server-5.7 as supported packages. Obviously rolled by the Ubuntu team, but fully supported for arm64.

**Comment by yanxg** [comment]
I solved this problem by using `image: mysql:8.0.26` and `platform: linux/x86_64` in my M1 (ARM64)

**Comment by dbwodlf3** [comment]
AWS t4 instances architecture is aarch64 but mysql:5.6 doesn't support aarch 64...

**Comment by emilioSp** [comment]
Supporting arm64 would be very appreciated

**Comment by adminy** [comment]
booo, this should have support! still an issue.

**Comment by andytson** [comment]
> booo, this should have support! still an issue.

It's closed because it's done, see `mysql:8.0-oracle`

**Comment by adminy** [comment]
Then why mysql:latest isn't working?

**Comment by andytson** [comment]
See https://hub.docker.com/_/mysql, tag `oracle` for the equivalent to `latest` for this.

It's another "flavour" of the image using oraclelinux (since Oracle only supply pre-compiled binaries for oraclelinux)
